README.md
# portainer_deploy_tool
使用portainer部署更新服务的工具


CHANGELOG.md
# 0.0.3

## 接口修改

+ 新增全局flag`not_verify`用于跳过https的验证
+ 子命令`updateservicebywebhook`的`tokens`参数现在是位置参数
+ 子命令`updateservicebywebhook`现在可以用执行目录下的`update_service_by_webhooks_config.json`文件作为配置
+ 子命令`createorupdatestack`的`endpoints`参数现在是位置参数
+ 子命令`createorupdatestack`现在可以用执行目录下的`create_or_update_stack_config.json`文件作为配置
+ 子命令`updateserviceinstack`的`deploy_path`参数参数现在是位置参数
+ 子命令`updateserviceinstack`现在可以用执行目录下的`update_service_in_stack_config.json`文件作为配置

## bug修复

+ 修复`updateservicebywebhook`的调用url错误

# 0.0.2

## 接口修改

+ `updateservicebywebhook`现在可以设置多个token
+ `artifact_version`和`tag_prefix`参数被取消
+ 现在`retry_max_times`和`retry_interval_backoff_factor`作为哦三个子命令的共有参数,`retry_interval`被移除,现在重试的间隔时间将根据公式`{backoff factor} * (2 ** ({number of total retries} - 1))`获得

# 0.0.1

项目创建

portainer_deploy_tool/utils.py
import requests
from pyloggerhelper import log

base_schema_properties = {
    "log_level": {
        "type": "string",
        "title": "l",
        "description": "log等级",
        "enum": ["DEBUG", "INFO", "WARN", "ERROR"],
        "default": "DEBUG"
    },
    "base_url": {
        "type": "string",
        "title": "b",
        "description": "portainer的根url"
    },
    "retry_max_times": {
        "type": "integer",
        "description": "重试次数",
    },
    "retry_interval_backoff_factor": {
        "type": "number",
        "description": "重试间隔时间,的参数,间隔时间位`{backoff factor} * (2 ** ({number of total retries} - 1))`",
        "default": 0.1
    },
    "not_verify": {
        "type": "boolean",
        "description": "是否不验证ssl",
        "default": False
    },
}


class HttpCodeError(Exception):
    """http请求返回错误"""
    pass


def get_jwt(rq: requests.Session, base_url: str, username: str, password: str) -> str:
    """获取jwt.

    Args:
        rq (requests.Session): 请求会话
        base_url (str): portainer的根地址
        username (str): portainer用户名
        password (str): 用户的密码

    Returns:
        str: jwt的值
    """
    res = rq.post(
        base_url + "/api/auth",
        json={
            "Username": username,
            "Password": password
        }
    )
    if res.status_code != 200:
        log.error("get jwt query get error",
                  base_url=base_url,
                  username=username,
                  status_code=res.status_code)
        raise HttpCodeError("get jwt query get error")
    try:
        res_json = res.json()
    except Exception as e:
        log.error("get jwt query get json result error",
                  base_url=base_url,
                  username=username,
                  err=type(e),
                  err_msg=str(e),
                  exc_info=True,
                  stack_info=True)
        raise e
    else:
        jwt = res_json.get("jwt")
        if jwt:
            return jwt
        else:
            log.error("get jwt query has no field jwt",
                      base_url=base_url,
                      username=username,
                      res_json=res_json)
            raise AttributeError("get jwt query has no field jwt")


portainer_deploy_tool/create_or_update_stack.py
from pathlib import Path
from dataclasses import dataclass
from copy import deepcopy
from collections import defaultdict
from typing import Dict, Any, List, NamedTuple, DefaultDict, Optional
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util import Retry
from schema_entry import EntryPoint
from pyloggerhelper import log
from .utils import (
    HttpCodeError,
    base_schema_properties,
    get_jwt
)


schema_properties: Dict[str, Any] = {
    "username": {
        "type": "string",
        "title": "u",
        "description": "portainer的登录用户名"
    },
    "password": {
        "type": "string",
        "title": "p",
        "description": "portainer的登录用户名对应的密码",
    },
    "endpoints": {
        "type": "array",
        "title": "d",
        "description": "指定要部署的节点",
        "items": {
            "type": "integer"
        }
    },
    "excepts": {
        "type": "array",
        "title": "e",
        "description": "指定部署节点上排除哪些stack不要进行操作,格式为`endpoint::stackname`.",
        "items": {
            "type": "string"
        }
    },
    "suffix": {
        "type": "string",
        "title": "s",
        "description": "用于匹配的文件文件名后缀,注意除去后缀外的部分会作为stack名,而stack名中不能有`.`",
        "default": ".docker-compose.yaml"
    },
    "prune": {
        "type": "boolean",
        "description": "更新时是否清除不用的资源",
        "default": False

    },
    "repository_reference_name": {
        "type": "string",
        "description": "分支信息",
        "default": "refs/heads/master"
    },
    "repository_url": {
        "type": "string",
        "description": "git仓库根路径"
    },
    "repository_username": {
        "type": "string",
        "description": "git仓库用户名",
    },
    "repository_password": {
        "type": "string",
        "description": "git仓库用户名对应的密码",
    },
    "cwd": {
        "type": "string",
        "description": "执行位置",
        "default": "."
    }
}

schema_properties.update(**base_schema_properties)


class NotSwarmEndpointError(Exception):
    """节点不是swarm节点"""
    pass


class ComposeInfo(NamedTuple):
    stack_name: str
    path: str


@dataclass
class GitStackInfo:
    endpoint_id: int
    stack_name: str
    stack_id: Optional[int] = None
    swarm_id: Optional[str] = None
    composeFilePathInRepository: Optional[str] = None
    repositoryReferenceName: Optional[str] = None
    repositoryURL: Optional[str] = None
    env: Optional[List[Dict[str, str]]] = None

    def check_git_repository(self) -> None:
        if not (self.composeFilePathInRepository and self.repositoryReferenceName and self.repositoryURL):
            raise AttributeError("need git repository info")

    def update_to_portainer(self, rq: requests.Session, base_url: str, jwt: str, prune: bool = False,
                            repositoryUsername: Optional[str] = None, repositoryPassword: Optional[str] = None) -> None:
        """更新portainer中的stack

        Args:
            base_url (str): portainer的根地址
            jwt (str): 访问jwt
            prune (bool, optional): 更新是否删除不再使用的资源. Defaults to False.
            repositoryUsername (Optional[str], optional): git仓库账户. Defaults to None.
            repositoryPassword (Optional[str], optional): git仓库密码. Defaults to None.

        Raises:
            HttpCodeError: update stack query get error
            e: update stack query get json result error
        """
        body: Dict[str, Any] = {
            "env": self.env,
            "prune": prune,
            "repositoryReferenceName": self.repositoryReferenceName,
        }
        if repositoryUsername and repositoryPassword:
            repositoryAuthentication = True
            body.update({
                "repositoryAuthentication": repositoryAuthentication,
                "repositoryPassword": repositoryPassword,
                "repositoryUsername": repositoryUsername
            })
        else:
            repositoryAuthentication = False
            body.update({
                "repositoryAuthentication": repositoryAuthentication
            })
        res = rq.put(
            f"{base_url}/api/stacks/{self.stack_id}/git",
            headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt}),
            params={
                "endpointId": self.endpoint_id
            },
            json=body)
        if res.status_code != 200:
            log.error("update stack query get error",
                      base_url=base_url,
                      status_code=res.status_code,
                      stack=self)
            raise HttpCodeError("update stack query get error")
        try:
            res_json = res.json()
        except Exception as e:
            log.error("update stack query get json result error", stack=self, err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
            raise e
        else:
            log.info("update stack ok", stack=self, result=res_json)

    def create_to_portainer(self, rq: requests.Session, base_url: str, jwt: str,
                            repositoryUsername: Optional[str] = None, repositoryPassword: Optional[str] = None) -> None:
        """使用对象的信息创建stack.

        Args:
            base_url (str):  portainer的根地址
            jwt (str): 访问jwt
            repositoryUsername (Optional[str], optional): git仓库账户. Defaults to None.
            repositoryPassword (Optional[str], optional): git仓库密码. Defaults to None.

        Raises:
            AttributeError: only create git repository method
            HttpCodeError: create stack query get error
            e: create stack query get json result error
        """
        body: Dict[str, Any] = {
            "env": self.env,
            "composeFilePathInRepository": self.composeFilePathInRepository,
            "name": self.stack_name,
            "repositoryReferenceName": self.repositoryReferenceName,
            "repositoryURL": self.repositoryURL,
        }
        if repositoryUsername and repositoryPassword:
            repositoryAuthentication = True
            body.update({
                "repositoryAuthentication": repositoryAuthentication,
                "repositoryPassword": repositoryPassword,
                "repositoryUsername": repositoryUsername
            })
        else:
            repositoryAuthentication = False
            body.update({
                "repositoryAuthentication": repositoryAuthentication
            })
        if self.swarm_id:
            stack_type = 1
            body.update({"swarmID": self.swarm_id})
        else:
            stack_type = 2
        params = (("method", "repository"), ("type", stack_type), ("endpointId", self.endpoint_id))
        res = rq.post(
            f"{base_url}/api/stacks",
            headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt}),
            params=params,
            json=body)
        if res.status_code != 200:
            log.error("create stack query get error",
                      base_url=base_url,
                      status_code=res.status_code,
                      stack=self)
            raise HttpCodeError("create stack query get error")
        try:
            res_json = res.json()
        except Exception as e:
            log.error("create stack query get json result error", stack=self, err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
            raise e
        else:
            log.info("create stack ok", stack=self, result=res_json)

    def update_or_create(self, rq: requests.Session, base_url: str, jwt: str, prune: bool = False,
                         repositoryUsername: Optional[str] = None, repositoryPassword: Optional[str] = None) -> None:
        self.check_git_repository()
        if self.stack_id:
            self.update_to_portainer(rq=rq, base_url=base_url, jwt=jwt, prune=prune, repositoryUsername=repositoryUsername, repositoryPassword=repositoryPassword)
        else:
            self.create_to_portainer(rq=rq, base_url=base_url, jwt=jwt, repositoryUsername=repositoryUsername, repositoryPassword=repositoryPassword)
        log.info("update_or_create query ok")


class CreateOrUpdateStack(EntryPoint):
    """扫描指定目录下的compose文件,在指定的端点中如果已经部署则更新stack,否则创建stack."""
    default_config_file_paths = [
        "./create_or_update_stack_config.json"
    ]
    argparse_noflag = "endpoints"
    schema = {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "object",
        "required": ["base_url", "username", "password", "endpoints", "suffix", "cwd", "repository_reference_name", "repository_url"],
        "properties": schema_properties
    }

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
        self.compose_infos: List[ComposeInfo] = []

    def find_compose(self, cwd: Path, suffix: str, root: Path) -> None:
        for p in cwd.iterdir():
            if p.is_file:
                if p.name.endswith(suffix):
                    rp = p.relative_to(root)
                    stack_name = p.name.replace(suffix, "")
                    if stack_name in [i.stack_name for i in self.compose_infos]:
                        log.error("stack name repetition", stack_name=stack_name)
                        raise AttributeError("stack name repetition")
                    ci = ComposeInfo(stack_name=stack_name, path=str(rp))
                    self.compose_infos.append(ci)
            elif p.is_dir():
                self._find_compose(cwd=p, suffix=suffix, root=root)

    def handdler_excepts(self, excepts: Optional[List[str]] = None) -> DefaultDict[int, List[ComposeInfo]]:
        result: DefaultDict[int, List[ComposeInfo]] = defaultdict(lambda: deepcopy(self.compose_infos))
        if excepts:
            temp = defaultdict(set)
            for exp in excepts:
                endpoint, stack_name = exp.split("::")
                temp[int(endpoint)].add(stack_name)
            for group, content in temp.items():
                total = deepcopy(self.compose_infos)
                result[group] = [i for i in total if i.stack_name not in content]
        return result

    def do_main(self) -> None:
        """入口程序."""
        base_url = self.config["base_url"]
        log_level = self.config["log_level"]
        username = self.config["username"]
        password = self.config["password"]
        endpoints = self.config["endpoints"]
        suffix = self.config["suffix"]
        cwd = self.config["cwd"]
        excepts = self.config.get("excepts")
        prune = self.config.get("prune")
        repository_url = self.config.get("repository_url")
        repository_reference_name = self.config.get("repository_reference_name")
        repository_username = self.config.get("repository_username")
        repository_password = self.config.get("repository_password")
        retry_max_times = self.config.get("retry_max_times")
        retry_interval_backoff_factor = self.config.get("retry_interval_backoff_factor")
        not_verify = retry_interval_backoff_factor = self.config.get("not_verify")
        if not_verify:
            rq = requests.Session(verify=False)
        else:
            rq = requests.Session()
        if retry_max_times and int(retry_max_times) > 0:
            rq.mount('https://', HTTPAdapter(max_retries=Retry(total=int(retry_max_times), backoff_factor=retry_interval_backoff_factor, method_whitelist=frozenset(['GET', 'POST', 'PUT']))))
        # 初始化log
        log.initialize_for_app(app_name="UpdateStack", log_level=log_level)
        log.info("get config", config=self.config)
        # 找到所有符合条件的stack
        cwdp = Path(cwd)
        root = Path(".")
        self.find_compose(cwd=cwdp, suffix=suffix, root=root)
        # 处理excepts
        endpoints_stacks = self.handdler_excepts(excepts)
        log.debug("deal with excepts ok", endpoints_stacks=endpoints_stacks)
        # 获取jwt
        jwt = get_jwt(rq, base_url=base_url, username=username, password=password)
        log.debug("deal with jwt ok", jwt=jwt)
        # 获取已经存在的stack信息
        endpoint_stack_info = get_all_stacks_from_portainer(rq, base_url=base_url, jwt=jwt)
        log.debug("deal with endpoint_stack_info ok", endpoint_stack_info=endpoint_stack_info)
        # 获取endpoint信息
        for endpoint in endpoints:
            swarmID: Optional[str] = None
            try:
                swarmID = get_swarm_id(rq, base_url=base_url, jwt=jwt, endpoint=endpoint)
            except NotSwarmEndpointError:
                log.info("Endpoint not swarm", endpoint=endpoint)
            except Exception as e:
                raise e
            log.debug("deal with swarmID ok", swarmID=swarmID, endpoint=endpoint)
            exist_stacks = endpoint_stack_info.get(endpoint)
            if exist_stacks is None:
                continue
            preparing_stacks = endpoints_stacks[endpoint]
            for _stack in preparing_stacks:
                stack = exist_stacks.get(_stack.stack_name)
                if stack:
                    if stack.repositoryURL == repository_url and stack.repositoryReferenceName == repository_reference_name and stack.composeFilePathInRepository == _stack.path and stack.swarm_id == swarmID:
                        stack.update_or_create(rq, base_url=base_url, jwt=jwt, prune=prune,
                                               repositoryUsername=repository_username, repositoryPassword=repository_password)
                else:
                    stack = GitStackInfo(
                        endpoint_id=endpoint,
                        stack_name=_stack.stack_name,
                        swarm_id=swarmID,
                        composeFilePathInRepository=_stack.path,
                        repositoryReferenceName=repository_reference_name,
                        repositoryURL=repository_url, env=[])
                    stack.update_or_create(rq, base_url=base_url, jwt=jwt, prune=prune,
                                           repositoryUsername=repository_username, repositoryPassword=repository_password)


def get_swarm_id(rq: requests.Session, base_url: str, jwt: str, endpoint: int) -> str:
    """获取端点的SwarmID.

    Args:
        rq (requests.Session): 请求会话
        base_url (str): portainer的根地址
        jwt (str): 访问jwt
        endpoint (int): 端点ID

    Raises:
        HttpCodeError: get swarm id query get error
        e: get swarm id query get json result error
        AssertionError: endpint not swarm

    Returns:
        str: Swarm ID
    """
    res = rq.get(f"{base_url}/api/endpoints/{endpoint}/docker/swarm",
                 headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt}))
    if res.status_code != 200:
        log.error("get swarm id query get error",
                  base_url=base_url,
                  endpoint=endpoint,
                  status_code=res.status_code)
        raise HttpCodeError("get swarm id query get error")
    try:
        res_json = res.json()
    except Exception as e:
        log.error("get swarm id query get json result error", endpoint=endpoint, err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
        raise e
    else:
        swarm_id = res_json.get("ID")
        if swarm_id:
            return swarm_id
        else:
            raise NotSwarmEndpointError(f"endpint {endpoint} not swarm")


def get_all_stacks_from_portainer(rq: requests.Session, base_url: str, jwt: str) -> Dict[int, Dict[str, GitStackInfo]]:
    """

    Args:
        rq (requests.Session): 请求会话
        base_url (str): portainer的根地址
        jwt (str): 访问jwt

    Raises:
        HttpCodeError: get stack query get error
        e: get stack query get json result error

    Returns:
        Dict[int, Dict[str, GitStackInfo]]: dict[endpointid,dict[stack_name,stackinfo]]
    """
    res = rq.get(f"{base_url}/api/stacks",
                 headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt}))
    if res.status_code != 200:
        log.error("get swarm id query get error",
                  base_url=base_url,
                  status_code=res.status_code)
        raise HttpCodeError("get stack query get error")
    try:
        res_jsons = res.json()
    except Exception as e:
        log.error("get stack query get json result error", err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
        raise e
    else:
        result: Dict[int, Dict[str, GitStackInfo]] = {}
        for res_json in res_jsons:
            gcf = res_json.get('GitConfig')
            endpoint_id = res_json['EndpointId']
            stack_id = res_json['Id']
            stack_name = res_json["Name"]
            if gcf:
                gsi = GitStackInfo(
                    endpoint_id=endpoint_id,
                    env=res_json["Env"],
                    stack_id=stack_id,
                    stack_name=stack_name,
                    swarm_id=res_json.get('SwarmId'),
                    composeFilePathInRepository=gcf["ConfigFilePath"],
                    repositoryReferenceName=gcf["ReferenceName"],
                    repositoryURL=gcf["URL"]
                )
            else:
                gsi = GitStackInfo(
                    endpoint_id=endpoint_id,
                    env=res_json["Env"],
                    stack_id=stack_id,
                    stack_name=res_json["Name"],
                    swarm_id=res_json.get('SwarmId'),
                    composeFilePathInRepository=None,
                    repositoryReferenceName=None,
                    repositoryURL=None
                )
            if not result.get(endpoint_id):
                result[endpoint_id] = {stack_name: gsi}
            else:
                result[endpoint_id][stack_name] = gsi

        return result


portainer_deploy_tool/update_service_by_webhook.py
from typing import Dict, Any
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util import Retry
from schema_entry import EntryPoint
from pyloggerhelper import log
from .utils import base_schema_properties, HttpCodeError


schema_properties: Dict[str, Any] = {
    "tokens": {
        "type": "array",
        "items": {
            "type": "string"
        },
        "title": "o",
        "description": "webhook的token"
    },
}

schema_properties.update(**base_schema_properties)


class UpdateServiceByWebhooks(EntryPoint):
    """利用webhook更新已经存在的service."""
    default_config_file_paths = [
        "./update_service_by_webhooks_config.json"
    ]
    argparse_noflag = "tokens"
    schema = {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "object",
        "required": ["base_url", "tokens"],
        "properties": schema_properties
    }

    def do_main(self) -> None:
        """入口程序."""
        base_url = self.config["base_url"]
        log_level = self.config["log_level"]
        tokens = self.config["tokens"]
        retry_max_times = self.config.get("retry_max_times")
        retry_interval_backoff_factor = self.config.get("retry_interval_backoff_factor", 0)
        not_verify = retry_interval_backoff_factor = self.config.get("not_verify")
        if not_verify:
            rq = requests.Session(verify=False)
        else:
            rq = requests.Session()
        if retry_max_times and int(retry_max_times) > 0:
            rq.mount('https://', HTTPAdapter(max_retries=Retry(total=int(retry_max_times), backoff_factor=retry_interval_backoff_factor, method_whitelist=frozenset(['GET', 'POST', 'PUT']))))
        log.initialize_for_app(app_name="UpdateStack", log_level=log_level)
        for token in tokens:
            res = rq.post(f"{base_url}/api/webhooks/{token}")
            if res.status_code >= 300 or res.status_code <= 199:
                log.error("update service query get error", status_code=res.status_code)
            else:
                log.info("update service query ok", status_code=res.status_code)


portainer_deploy_tool/update_service_in_stack.py
from typing import Dict, Any, List, Optional
import yaml
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util import Retry
from schema_entry import EntryPoint
from pyloggerhelper import log
from .utils import base_schema_properties, get_jwt, HttpCodeError


schema_properties: Dict[str, Any] = {
    "username": {
        "type": "string",
        "title": "u",
        "description": "portainer的登录用户名"
    },
    "password": {
        "type": "string",
        "title": "p",
        "description": "portainer的登录用户名对应的密码",
    },
    "registry_name": {
        "type": "string",
        "title": "n",
        "description": "镜像仓库名",
    },
    "artifact_name": {
        "type": "string",
        "title": "n",
        "description": "要更新的镜像名",
    },
    "artifact_version": {
        "type": "string",
        "title": "v",
        "description": "要更新的镜像版本",
        "default": "latest"
    },
    "tag_prefix": {
        "type": "string",
        "title": "t",
        "description": "待更新路径"
    },
    "deploy_path": {
        "type": "array",
        "items": {
            "type": "string"
        },
        "title": "d",
        "description": "待更新路径,形式为`[ENDPOINT_ID]/[STACK_ID]/[SERVICE_NAME]`"
    }
}

schema_properties.update(**base_schema_properties)


class UpdateServiceInStack(EntryPoint):
    """更新已经存在的stack."""
    default_config_file_paths = [
        "./update_service_in_stack_config.json"
    ]
    argparse_noflag = "deploy_path"
    schema = {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "object",
        "required": ["base_url", "username", "password", "artifact_name", "artifact_version", "deploy_path"],
        "properties": schema_properties
    }

    def parser_deploy_path(self, deploy_path: List[str]) -> Dict[str, List[str]]:
        """解析deploy_path便于部署.

        Args:
            deploy_path (List[str]): 待更新路径,形式为`[ENDPOINT_ID]/[STACK_ID]/[SERVICE_NAME],[ENDPOINT_ID]/[STACK_ID]/[SERVICE_NAME]...`

        Raises:
            AttributeError: stack_id must be digit
            AttributeError: endpoint_id must be digit

        Returns:
            Dict[str, List[str]]: key为<endpoint_id_str>::<stack_id>;value为其中要更新的service名称列表
        """
        services: Dict[str, List[str]] = {}
        for p in deploy_path:
            endpoint_id, stack_id, service_name = p.split("/")
            if not stack_id.isdigit():
                log.error("stack_id must be digit", stack_id=stack_id)
                raise AttributeError(f"stack_id must be digit, but is {stack_id}")
            if not endpoint_id.isdigit():
                log.error("endpoint_id must be digit", endpoint_id=endpoint_id)
                raise AttributeError(f"endpoint_id must be digit, but is{endpoint_id}")

            stack_key = f"{endpoint_id}::{stack_id}"
            if services.get(stack_key):
                services[stack_key].append(service_name)
            else:
                services[stack_key] = [service_name]
        return services

    def make_image_name(self, artifact_name: str, artifact_version: str, *,
                        registry_name: Optional[str] = None, tag_prefix: Optional[str] = None) -> str:
        """[summary]

        Args:
            artifact_name (str): 制品名
            artifact_version (str): 制品版本
            registry_name (Optional[str], optional): 镜像仓库hostname. Defaults to None.
            tag_prefix (Optional[str], optional): 制品版本标签前缀. Defaults to None.

        Returns:
            str: 完整镜像名
        """
        if registry_name:
            if tag_prefix:
                image_name = f"{registry_name}/{artifact_name}:{tag_prefix}-{artifact_version}"
            else:
                image_name = f"{registry_name}/{artifact_name}:{artifact_version}"
        else:
            if tag_prefix:
                image_name = f"{artifact_name}:{tag_prefix}-{artifact_version}"
            else:
                image_name = f"{artifact_name}:{artifact_version}"
        return image_name

    def get_stack_file_content(self, rq: requests.Session, base_url: str, jwt: str, stack_id: str) -> str:
        """获取更新前stack的compose文本.

        Args:
            rq (requests.Session): 请求会话
            base_url (str): portainer基础路径
            jwt (str): 访问jwt
            stack_id (str): stack信息的id

        Raises:
            HttpCodeError: get stack file content query get error
            e: get stack file content query get json result error
            AttributeError: get stack file content query has no field StackFileContent

        Returns:
            str: 旧compose文本
        """
        res = rq.get(
            f"{base_url}/stacks/{stack_id}/file",
            headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt})
        )
        if res.status_code != 200:
            log.error("get stack file content query get error", stack_id=stack_id, status_code=res.status_code)
            raise HttpCodeError("get stack file content query get error")
        try:
            res_json = res.json()
        except Exception as e:
            log.error("get stack file content query get json result error", stack_id=stack_id, err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
            raise e
        else:
            StackFileContent = res_json.get("StackFileContent")
            if StackFileContent:
                return StackFileContent
            else:
                log.error("get stack file content query has no field StackFileContent", stack_id=stack_id)
                raise AttributeError("get stack file content query has no field StackFileContent")

    def deploy(self, rq: requests.Session, base_url: str, jwt: str, image_name: str, stack_key: str, services: List[str]) -> None:
        """更新部署单个stack中的服务.

        Args:
            rq (requests.Session): 请求会话
            base_url (str): portainer基础路径
            jwt (str): 访问jwt
            image_name (str): 镜像完整名
            stack_key (str): stack信息,形式为<endpoint_id_str>::<stack_id>
            services (List[str]): 这个stack中要更新的services名称列表

        Raises:
            HttpCodeError: service_name not in old compose file
            AttributeError: deploy query get error
            e: deploy query get json result error
        """
        endpoint_id_str, stack_id = stack_key.split("::")
        endpoint_id = int(endpoint_id_str)
        StackFileContent = self.get_stack_file_content(rq, base_url=base_url, jwt=jwt, stack_id=stack_id)
        s = yaml.load(StackFileContent)
        log.info("get old compose", content=s)
        for service_name in services:
            if s['services'].get(service_name):
                s['services'][service_name]['image'] = image_name
            else:
                log.error("service_name not in old compose file", service_name=service_name)
                raise AttributeError(f"service_name {service_name} not in old compose file")
        compose = yaml.dump(s, sort_keys=False)
        res = rq.put(
            f"{base_url}/stacks/{stack_id}",
            headers=requests.structures.CaseInsensitiveDict({"Authorization": "Bearer " + jwt}),
            params={"endpointId": endpoint_id},
            json={
                "StackFileContent": compose,
                "Prune": False
            }
        )
        if res.status_code != 200:
            log.error("deploy query get error", stack_key=stack_key, status_code=res.status_code)
            raise HttpCodeError("deploy query get error")
        else:
            try:
                res_json = res.json()
            except Exception as e:
                log.error("deploy query get json result error", stack_key=stack_key, err=type(e), err_msg=str(e), exc_info=True, stack_info=True)
                raise e
            else:
                log.debug("deploy query get result", stack_key=stack_key, cotent=res_json)
                log.info("deploy ok", stack_key=stack_key)

    def deploy_all(self, rq: requests.Session, base_url: str, jwt: str, image_name: str, stack_services: Dict[str, List[str]]) -> None:
        """部署解析出来的所有stack.

        Args:
            rq (requests.Session): 请求会话
            base_url (str): portainer基础路径
            jwt (str): 访问jwt
            image_name (str): 镜像完整名
            stack_services (List[str]): 解析出来的stack-service信息
        """
        for stack_key, services in stack_services.items():
            self.deploy(
                rq,
                base_url=base_url,
                jwt=jwt,
                image_name=image_name,
                stack_key=stack_key,
                services=services)
        log.info("deploy all done")

    def do_main(self) -> None:
        """入口程序."""
        base_url = self.config["base_url"]
        log_level = self.config["log_level"]
        username = self.config["username"]
        password = self.config["password"]
        artifact_name = self.config["artifact_name"]
        artifact_version = self.config["artifact_version"]
        deploy_path = self.config["deploy_path"]
        retry_max_times = self.config.get("retry_max_times")
        retry_interval_backoff_factor = self.config.get("retry_interval_backoff_factor")
        not_verify = retry_interval_backoff_factor = self.config.get("not_verify")
        if not_verify:
            rq = requests.Session(verify=False)
        else:
            rq = requests.Session()
        if retry_max_times and int(retry_max_times) > 0:
            rq.mount('https://', HTTPAdapter(max_retries=Retry(total=int(retry_max_times), backoff_factor=retry_interval_backoff_factor, method_whitelist=frozenset(['GET', 'POST', 'PUT']))))
        log.initialize_for_app(app_name="UpdateStack", log_level=log_level)
        jwt = get_jwt(rq, base_url=base_url, username=username, password=password)
        image_name = self.make_image_name(
            artifact_name=artifact_name,
            artifact_version=artifact_version,
            registry_name=self.config.get("registry_name"),
            tag_prefix=self.config.get("tag_prefix")
        )
        stack_services = self.parser_deploy_path(deploy_path)
        self.deploy_all(
            rq,
            base_url=base_url,
            jwt=jwt,
            image_name=image_name,
            stack_services=stack_services
        )
